### ISSUE TITLE
New lint: replace `iter.filter(x).count() > 0` with `iter.any(x)`

### ISSUE THREAD

**Issue by ebroto** [rust-lang/rust-clippy#5932]
### What it does

It suggests replacing `iter.filter(x).count() > 0` with `iter.any(x)`. The suggestion should be `MachineApplicable`.

### Categories (optional)

- Kind: complexity

*What is the advantage of the recommended code over the original code*

It's more concise.

### Drawbacks

None.

### Example

```rust
segments.iter().filter(|ps| ps.ident.as_str() == "prelude").count() > 0
```

Could be written as:

```rust
segments.iter().any(|ps| ps.ident.as_str() == "prelude")
```

**Comment by matthiaskrgr** [comment]
I think this might as well fit into the `perf` category since `.filter().count()` has to walk through all iterator items but `.any()` can return early?

**Comment by matthiaskrgr** [comment]
Hmm, but that might also cause false positives for the case that the `.filter()` call has side effects and the `.any()` call reduces the number of side-effects that happen :/

**Comment by ebroto** [comment]
Indeed! That's #3351

**Comment by flip1995** [comment]
The lint idea is good, but we have to think about side-effects. I think this is hard to do/to get right. Cases where we have side effects would be:

- mutating an `UpVar`
- calling a function, which has side effects (hiding this inside an iterator is a bad idea IMO, but suggesting replacing this with `any` would still change the semantics)
- Any kind of I/O
- ???

**Comment by camsteffen** [comment]
This could be generalized to suggest `iter.count() > 0` -> `iter.next().is_some()` in some cases.

**Comment by pickfire** [comment]
`all()` could also be applied here as well as `!= 0` rather than `> 0`.
